enterprise/server/crypter_service/crypter_service.go

```diff
@@ -0,0 +1,291 @@
+	plainTextChunkSize           = 1024 * 1024 // 1 MiB
+	// 4 byte chunk counter, 24 bytes group ID
+	encryptedChunkAuthHeaderSize = 28
+	encryptedChunkOverhead       = gcmStandardNonceSizeBytes + gcmTagSize
+)
+
+// TODO(vadim): pool buffers to reduce allocations
+// TODO(vadim): include file digest in authentication header
+type Crypter struct {
+	auth interfaces.Authenticator
+	dbh  interfaces.DBHandle
+	kms  interfaces.KMS
+}
+
+func New(env environment.Env) *Crypter {
+	return &Crypter{
+		kms:  env.GetKMS(),
+		auth: env.GetAuthenticator(),
+		dbh:  env.GetDBHandle(),
+	}
+}
+
+type Encryptor struct {
+	key           *tables.EncryptionKeyVersion
+	ciph          cipher.AEAD
+	groupIDHeader []byte
+	w             interfaces.CommittedWriteCloser
+	wroteHeader   bool
+	chunkCounter  uint32
+	nonceBuf      []byte
+
+	// buf collects the plaintext until there's enough for a full chunk or the
+	// is no more data left to encrypt.
+	buf []byte
+	// bufIdx is the index into the buffer where new data should be written.
+	bufIdx int
+	// bufCap is the maximum amount of plaintext the buffer can hold. The raw
+	// buffer is larger to allow encryption to be done in place.
+	bufCap int
+}
+
+func fixedWidthGroupID(groupID string) []byte {
+	return []byte(fmt.Sprintf("%24s", groupID))
+}
+
+func makeChunkAuthHeader(chunkIndex uint32, groupIDHeader []byte) ([]byte, error) {
+	if 4+len(groupIDHeader) != encryptedChunkAuthHeaderSize {
+		return nil, status.FailedPreconditionError("invalid header size")
+	}
+
+	chunkHeader := make([]byte, encryptedChunkAuthHeaderSize)
+	binary.LittleEndian.PutUint32(chunkHeader, chunkIndex)
+	copy(chunkHeader[4:], groupIDHeader)
+	return chunkHeader, nil
+}
+
+func (e *Encryptor) flushBlock() error {
+	if e.bufIdx == 0 {
+		return nil
+	}
+
+	if _, err := rand.Read(e.nonceBuf); err != nil {
+		return err
+	}
+	if _, err := e.w.Write(e.nonceBuf); err != nil {
+		return err
+	}
+
+	e.chunkCounter++
+	chunkAuth, err := makeChunkAuthHeader(e.chunkCounter, e.groupIDHeader)
+	if err != nil {
+		return err
+	}
+	ct := e.ciph.Seal(e.buf[:0], e.nonceBuf, e.buf[:e.bufIdx], chunkAuth)
+	if _, err := e.w.Write(ct); err != nil {
+		return err
+	}
+	e.bufIdx = 0
+	return nil
+}
+
+func (e *Encryptor) Write(p []byte) (n int, err error) {
+	if !e.wroteHeader {
+		if _, err := e.w.Write([]byte(encryptedDataHeaderSignature)); err != nil {
+			return 0, err
+		}
+		if _, err := e.w.Write([]byte{encryptedDataHeaderVersion}); err != nil {
+			return 0, err
+		}
+		e.wroteHeader = true
+	}
+
+	readIdx := 0
+	for readIdx < len(p) {
+		readLen := e.bufCap - e.bufIdx
+		if readLen > len(p)-readIdx {
+			readLen = len(p) - readIdx
+		}
+		copy(e.buf[e.bufIdx:], p[readIdx:readIdx+readLen])
+		e.bufIdx += readLen
+		readIdx += readLen
+		if e.bufIdx == e.bufCap {
+			if err := e.flushBlock(); err != nil {
+				return 0, err
+			}
+		}
+	}
+
+	return len(p), nil
+}
+
+func (e *Encryptor) Commit() error {
+	if err := e.flushBlock(); err != nil {
```

Comment:
Should we move flushing the final block to Write()? It's kind of confusing that Write() will not flush the last block if the bufCap is not a multiple of len(p)

i.e. If we hit the case where `readLen > len(p)-readIdx`, we won't enter the if statement `if e.bufIdx == e.bufCap` and flush the last block in Write

I could see it being confusing that all the blocks are flushed in Write() except the last one. i.e. If you're trying to debug a Write call and can't figure out why the last block keeps getting omitted